File: src/0791-Custom-Sort-String/custom_sort_string.py
from typing import Dict

class Solution:
    def customSortString(self, order: str, s: str) -> str:
        occ: Dict[str, int] = {}
        ans: str = ""

        for c in order:
            occ[c] = 0

        for c in s:
            if c in occ:
                occ[c] += 1

        for c, i in occ.items():
            ans += c * i
        
        for c in s:
            if c not in order:
                ans += c
        
        return ans


if __name__ == "__main__":
    order: str = "hucw"
    s: str = "utzoampdgkalexslxoqfkdjoczajxtuhqyxvlfatmptqdsochtdzgypsfkgqwbgqbcamdqnqztaqhqanirikahtmalzqjjxtqfnh"
    solution = Solution()

    print(solution.customSortString(order, s))

File: src/0791-Custom-Sort-String/custom_sort_string.cpp
#include <iostream>
#include <map>
#include <string>

class Solution {
public:
    std::string customSortString(std::string order, std::string s) {
        std::map<char, int> occ;
        std::string ans = "";

        for (char c : s) {
            occ[c]++;
        }

        for (char c : order) {
            if (occ.find(c) != occ.end()) {
                ans += std::string(occ[c], c);
                occ.erase(c);
            }
        }

        for (const auto& entry : occ) {
            ans += std::string(entry.second, entry.first);
        }

        return ans;
    }
};

int main() {
    std::string order = "hucw";
    std::string s = "utzoampdgkalexslxoqfkdjoczajxtuhqyxvlfatmptqdsochtdzgypsfkgqwbgqbcamdqnqztaqhqanirikahtmalzqjjxtqfnh";
    Solution solution;

    std::cout << solution.customSortString(order, s);

    return 0;
}